torchrl/modules/tensordict_module/rnn.py

```diff
@@ -0,0 +1,234 @@
+      multi-step. This class enables both usages.
+
+
+    After construction, the module is *not* set in temporal mode, ie. it will
+    expect single steps inputs.
+
+    If in temporal mode, it is expected that the last dimension of the tensordict
+    marks the number of steps. There is no constrain on the dimensionality of the
+    tensordict (except that it must be greater than one for temporal inputs).
+
+    Args:
+        lstm (torch.nn.Module): a :class:`torch.nn.LSTM` instance. For simplicity, this class does not
+            handle the construction of the module.
+        in_keys (list of str): a triplet of strings corresponding to the input value,
+            first and second hidden key.
+        out_keys (list of str): a triplet of strings corresponding to the output value,
+            first and second hidden key.
+            .. note::
+              For a better integration with TorchRL's environments, the best naming
+              for the output hidden key is ``("next", "hidden0")`` or similar, such
+              that the hidden values are passed from step to step during a rollout.
+
+    Attributes:
+        temporal_mode: Returns the temporal mode of the module.
+
+    Methods:
+        set_temporal_mode: controls whether the module should be executed in
+            temporal mode.
+
+    Examples:
+        >>> from torchrl.envs import TransformedEnv, InitTracker
+        >>> from torchrl.envs.libs.gym import GymEnv
+        >>> from torchrl.modules import MLP
+        >>> from torch import nn
+        >>> from tensordict.nn import TensorDictSequential as Seq, TensorDictModule as Mod
+        >>> env = TransformedEnv(GymEnv("Pendulum-v1"), InitTracker())
+        >>> lstm = nn.LSTM(input_size=env.observation_spec["observation"].shape[-1], hidden_size=64, batch_first=True)
+        >>> lstm_module = LSTMModule(lstm, in_keys=["observation", "hidden0", "hidden1"], out_keys=["intermediate", ("next", "hidden0"), ("next", "hidden1")])
+        >>> mlp = MLP(num_cells=[64], out_features=1)
+        >>> policy = Seq(lstm_module, Mod(mlp, in_keys=["intermediate"], out_keys=["action"]))
+        >>> policy(env.reset())
+        TensorDict(
+            fields={
+                action: Tensor(shape=torch.Size([1]), device=cpu, dtype=torch.float32, is_shared=False),
+                done: Tensor(shape=torch.Size([1]), device=cpu, dtype=torch.bool, is_shared=False),
+                intermediate: Tensor(shape=torch.Size([64]), device=cpu, dtype=torch.float32, is_shared=False),
+                is_init: Tensor(shape=torch.Size([1]), device=cpu, dtype=torch.bool, is_shared=False),
+                next: TensorDict(
+                    fields={
+                        hidden0: Tensor(shape=torch.Size([1, 64]), device=cpu, dtype=torch.float32, is_shared=False),
+                        hidden1: Tensor(shape=torch.Size([1, 64]), device=cpu, dtype=torch.float32, is_shared=False)},
+                    batch_size=torch.Size([]),
```

Comment:
Isn't it counterintuitive that 
obs -> intermediate -> action are all at root level but the hidden state is in next?

What again is the reason for this?

Reply:
the hidden is not used at time t but t+1 (obs, intermediate and action are for t and t only). It's probably a bit cleaner to have it in next as this is where it belongs... 

so you pack hidden in next such that you find it back when calling step_mdp
Per se we could avoid that, since I think most calls to step_mdp have keep_other=True (it is the default, it is the case for env.rollout and it is the case for the collectors!).
But if someone by mistake calls step_mdp with keep_other=False she'll loose the hidden, and since we populate the hidden with 0s when not found no error will be raised (outch!). 

So i'd rather avoid that scenario as much as I can!

Now i'm not against the idea of having it out of "next" if you find it's clearer.

Reply:
I see. So we put it in next because we know for sure that a step_mdp will happen.

Reply:
in the algo yes but it's your choice...

Reply:
I have to double-check, but I think RNNs are usually written as

`y_t, s_t = rnn(x_t, s_{t-1})`

in books/papers. I agree that perhaps `prev_state, state` makes more sense than `state, next_state`.

Reply:
There's what papers say and how it aligns with the rest of the API.
ie. I'm not sure how we could mark the input as being "prev" and the current as being "current" if the only containers we have are "current" and "next". Maybe a proper doc that mirrors what you're saying here?